src/main/java/org/jenkinsci/plugins/workflow/support/storage/LumpFlowNodeStorage.java

```diff
@@ -0,0 +1,271 @@
+                    if (roughNodes == null) {
+                        throw new IOException("Unable to load nodes, invalid data");
+                    }
+                    for (Tag t : roughNodes.values()) {
+                        FlowNode fn = t.node;
+                        try {
+                            FlowNode$exec.set(fn, exec);
+                        } catch (IllegalAccessException e) {
+                            throw (IllegalAccessError) new IllegalAccessError("Failed to set owner").initCause(e);
+                        }
+                        t.storeActions();
+                        for (FlowNodeAction a : Util.filter(t.actions(), FlowNodeAction.class)) {
+                            a.onLoad(fn);
+                        }
+                    }
+                    nodes = roughNodes;
+                } else {
+                    nodes = new HashMap<String, Tag>();
+                }
+            } else {
+                IOUtils.mkdirs(dir);
+                nodes = new HashMap<String, Tag>();
+            }
+        }
+        return nodes;
+    }
+
+    @Override
+    @CheckForNull
+    public FlowNode getNode(@Nonnull String id) throws IOException {
+        Tag t = getOrLoadNodes().get(id);
+        return (t != null) ? t.node : null;
+    }
+
+    public void storeNode(@Nonnull FlowNode n, boolean delayWritingActions) throws IOException {
+        Tag t = getOrLoadNodes().get(n.getId());
+        if (t != null) {
+            t.node = n;
+            List<Action> act = n.getActions();
+            t.actions = act.toArray(new Action[act.size()]);
+        } else {
+            getOrLoadNodes().put(n.getId(), new Tag(n, n.getActions()));
+        }
+        isModified = true;
+        if (!delayWritingActions) {
+            flush();
+        }
+    }
+
+    @Override
+    public void storeNode(FlowNode n) throws IOException {
+        storeNode(n, false);
+    }
+
+    /**
+     * Persists a single FlowNode to disk (if not already persisted).
+     * @param n Node to persist
+     * @throws IOException
+     */
+    @Override
+    public void flushNode(@Nonnull FlowNode n) throws IOException {
+        storeNode(n, false);
+    }
+
+    /** Force persisting any nodes that had writing deferred */
+    @Override
+    public void flush() throws IOException {
+        if (nodes != null && isModified) {
+            // TODO reuse a single buffer if we can, and consider using async FileChannel operations.
+            if (!dir.exists()) {
+                IOUtils.mkdirs(dir);
+            }
+            OutputStream os = new BufferedOutputStream(
+                    Files.newOutputStream(getStoreFile().toPath(), StandardOpenOption.WRITE, StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING)
+            );
+            XSTREAM.toXMLUTF8(nodes, os); // Hah, no atomic nonsense, just write and write and write!
```

Comment:
@jtnord Yes, it will error on load -- this is for cases where the user has explicitly chosen to prioritize performance over absolute integrity (for that build). 

In practice, my suspicion is that this will be a rather rare event -- it requires interruption during a fairly brief write operation that is performed only a few times per build.  Also there's downstream testcases to verify that we handle that case cleanly.

Reply:
ack.